src/Action.hs

```diff
@@ -164,75 +221,61 @@ fakeCookieTVar = unsafePerformIO (newTVarIO Nothing)
 -- FUTUREWORK: Move action implementation to another module and hide behind
 -- an API, similarly as it's done with persistent implementation,
 -- to reveal and mark (and possibly fix) where the implementation is hardwired.
-newtype Action r a = Action (ExceptT ActionExcept (RWST (r :~> IO) () UserState IO) a)
+newtype Action r a = Action (ExceptT ActionExcept (RWST (ActionEnv r) () UserState IO) a)
     deriving ( Functor
              , Applicative
              , Monad
              , MonadError ActionExcept
-             , MonadReader (r :~> IO)
+             , MonadReader (ActionEnv r)
              , MonadState UserState
              )
 
--- | Top level errors can happen.
---
--- FIXME: Create a different type
-type ActionExcept = ServantErr
-
 -- | Creates a natural transformation from Action to the servant handler monad.
---
--- FIXME:
--- - The ability to change the state is missing. FIXME: which state? login modifes UserState all right
--- - The state should be available after run. FIXME: which state? what for?
-mkRunAction :: forall r. PersistM r
-            => (r :~> IO) -> Action r :~> ExceptT ServantErr IO
-mkRunAction persistNat = Nat run
+-- See Frontend.runFrontend for the persistency of @UserState@.
+mkRunAction :: PersistM r => ActionEnv r -> Action r :~> ExceptT ServantErr IO
+mkRunAction env = Nat run
   where
-    run = ExceptT . fmap (view _1)
-        . runRWSTflip persistNat UserLoggedOut . runExceptT . (setCurrentUser >>) . unAction
+    run = withExceptT unActionExcept . ExceptT . fmap (view _1) . runRWSTflip env userLoggedOut
+        . runExceptT . unAction . (checkCurrentUser >>)
     unAction (Action a) = a
     runRWSTflip r s comp = runRWST comp r s
 
-    setCurrentUser :: ExceptT ActionExcept (RWST (r :~> IO) () UserState IO) ()
-    setCurrentUser = do
-      mcurrentUID <- liftIO . atomically $ readTVar fakeCookieTVar
-      uState <- case mcurrentUID of
-         Nothing -> return UserLoggedOut
-         Just uid -> do
-           muser <- liftIO . unNat persistNat $ findUser uid
-           let uLogin = case muser of
-                 Nothing -> error "mkRunAction: currently logged in user not in DB"
-                 Just user -> user ^. userLogin
-           return $ UserLoggedIn uLogin "session"
-      put uState
+    checkCurrentUser = do
+        isValid <- userState $ to validUserState
+        unless isValid $ do
+            logout
+            throwError500 "Invalid internal user session state"
```

Comment:
Isn't this status code <500?  It can be caused by invalid cookies, right?  (On the other hand, once the cookie has integrity protection, it wouldn't be triggered here.)